File: Android/shiba/src/main/java/moe/tlaster/shiba/converters/ShibaConverter.kt
package moe.tlaster.shiba.converters

import moe.tlaster.shiba.ShibaFunction
import moe.tlaster.shiba.ShibaFunctionExecutor
import moe.tlaster.shiba.SingleBindingShibaFunctionExecutor
import moe.tlaster.shiba.Singleton

abstract class ShibaConverter {
    abstract fun convert(value: Any?, parameter: Any?): Any?
}

class RawConverter : ShibaConverter() {
    override fun convert(value: Any?, parameter: Any?): Any? {
        return value
    }
}

open class FunctionConverter : ShibaConverter() {
    open val executor = Singleton.get<ShibaFunctionExecutor>()

    override fun convert(value: Any?, parameter: Any?): Any? {
        return when (parameter) {
            is ShibaFunction -> {
                executor.execute(parameter, value)
            }
            is ShibaConverterParameter -> {
                if (parameter.innerConverter != null) {
                    val innerResult = parameter.innerConverter.convert(value, parameter.innerParameter)
                    executor.execute(parameter.function, innerResult)
                } else {
                    executor.execute(parameter.function, value)
                }
            }
            else -> throw NotImplementedError()

        }
    }
}

class SingleBindingFunctionConverter : FunctionConverter() {
    override val executor = Singleton.get<SingleBindingShibaFunctionExecutor>()
}

internal data class ShibaConverterParameter(
        val innerConverter: ShibaConverter?,
        val innerParameter: Any?,
        val function: ShibaFunction
)

File: iOS/Shiba/Shiba/Convcerters/ShibaConverter.swift
//
// Created by 高垣朝陽 on 2018/8/16.
// Copyright (c) 2018 Tlaster. All rights reserved.
//

import Foundation

public class ShibaConverter {
    func convert(value: Any?, parameter: Any?) -> Any? {
        fatalError("convert(value:parameters:) has not been implemented")
    }

    static let Raw = RawConverter()
    static let Function = FunctionConverter()
    static let SingleBindingFunction = SingleBindingFunctionConverter()
}

public class RawConverter: ShibaConverter {
    override func convert(value: Any?, parameter: Any?) -> Any? {
        return value
    }
}

public class FunctionConverter: ShibaConverter {
    private let executor = ShibaFunctionExecutor()

    func getExecutor() -> ShibaFunctionExecutor {
        return executor
    }

    override func convert(value: Any?, parameter: Any?) -> Any? {
        if parameter is ShibaFunction {
            return getExecutor().execute(function: parameter as! ShibaFunction, dataContext: value)
        }
        fatalError()
    }
}

public class SingleBindingFunctionConverter: FunctionConverter {
    private let executor = SingleBindingShibaFunctionExecutor()

    override func getExecutor() -> ShibaFunctionExecutor {
        return executor;
    }
}
